File: psicogest/backend/app/api/v1/indicators.py
"""Therapy indicators router."""
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status

from app.core.deps import get_tenant_db, TenantDB
from app.schemas.therapy_indicator import (
    TherapyIndicatorCreate,
    TherapyIndicatorDetail,
    TherapyIndicatorUpdate,
    TherapyIndicatorWithMeasurements,
    TherapyMeasurementCreate,
    TherapyMeasurementDetail,
)
from app.services.therapy_indicator_service import IndicatorNotFoundError, TherapyIndicatorService

router = APIRouter(tags=["indicators"])


def _svc(ctx: TenantDB) -> TherapyIndicatorService:
    return TherapyIndicatorService(ctx.db, ctx.tenant.tenant_id)


@router.get("/patients/{patient_id}/indicators", response_model=list[TherapyIndicatorDetail])
def list_indicators(
    patient_id: uuid.UUID,
    ctx: Annotated[TenantDB, Depends(get_tenant_db)],
):
    indicators = _svc(ctx).list_by_patient(patient_id)
    return [TherapyIndicatorDetail.model_validate(indicator) for indicator in indicators]


@router.post(
    "/patients/{patient_id}/indicators",
    response_model=TherapyIndicatorDetail,
    status_code=status.HTTP_201_CREATED,
)
def create_indicator(
    patient_id: uuid.UUID,
    body: TherapyIndicatorCreate,
    ctx: Annotated[TenantDB, Depends(get_tenant_db)],
):
    indicator = _svc(ctx).create(patient_id, body)
    return TherapyIndicatorDetail.model_validate(indicator)


@router.get("/indicators/{indicator_id}", response_model=TherapyIndicatorWithMeasurements)
def get_indicator(
    indicator_id: uuid.UUID,
    ctx: Annotated[TenantDB, Depends(get_tenant_db)],
):
    try:
        indicator, measurements = _svc(ctx).get_with_measurements(indicator_id)
    except IndicatorNotFoundError:
        raise HTTPException(status_code=404, detail="Indicator not found")
    result = TherapyIndicatorWithMeasurements.model_validate(indicator)
    result.measurements = [TherapyMeasurementDetail.model_validate(m) for m in measurements]
    return result


@router.put("/indicators/{indicator_id}", response_model=TherapyIndicatorDetail)
def update_indicator(
    indicator_id: uuid.UUID,
    body: TherapyIndicatorUpdate,
    ctx: Annotated[TenantDB, Depends(get_tenant_db)],
):
    try:
        indicator = _svc(ctx).update(indicator_id, body)
        return TherapyIndicatorDetail.model_validate(indicator)
    except IndicatorNotFoundError:
        raise HTTPException(status_code=404, detail="Indicator not found")


@router.delete("/indicators/{indicator_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_indicator(
    indicator_id: uuid.UUID,
    ctx: Annotated[TenantDB, Depends(get_tenant_db)],
):
    try:
        _svc(ctx).delete(indicator_id)
    except IndicatorNotFoundError:
        raise HTTPException(status_code=404, detail="Indicator not found")


@router.post(
    "/indicators/{indicator_id}/measurements",
    response_model=TherapyMeasurementDetail,
    status_code=status.HTTP_201_CREATED,
)
def add_measurement(
    indicator_id: uuid.UUID,
    body: TherapyMeasurementCreate,
    ctx: Annotated[TenantDB, Depends(get_tenant_db)],
):
    try:
        measurement = _svc(ctx).add_measurement(indicator_id, body)
        return TherapyMeasurementDetail.model_validate(measurement)
    except IndicatorNotFoundError:
        raise HTTPException(status_code=404, detail="Indicator not found")


@router.get("/indicators/{indicator_id}/measurements", response_model=list[TherapyMeasurementDetail])
def list_measurements(
    indicator_id: uuid.UUID,
    ctx: Annotated[TenantDB, Depends(get_tenant_db)],
):
    try:
        measurements = _svc(ctx).list_measurements(indicator_id)
        return [TherapyMeasurementDetail.model_validate(measurement) for measurement in measurements]
    except IndicatorNotFoundError:
        raise HTTPException(status_code=404, detail="Indicator not found")

File: psicogest/frontend/src/lib/api/indicators.ts
import { request } from "./client";

export interface TherapyIndicator {
  id: string;
  patient_id: string;
  name: string;
  description: string | null;
  unit: string | null;
  initial_value: number | null;
  target_value: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TherapyMeasurement {
  id: string;
  indicator_id: string;
  session_id: string | null;
  value: number;
  notes: string | null;
  measured_at: string;
  created_at: string;
}

export interface TherapyIndicatorWithMeasurements extends TherapyIndicator {
  measurements: TherapyMeasurement[];
}

export interface TherapyIndicatorCreate {
  name: string;
  description?: string | null;
  unit?: string | null;
  initial_value?: number | null;
  target_value?: number | null;
}

export interface TherapyIndicatorUpdate extends Partial<TherapyIndicatorCreate> {
  is_active?: boolean;
}

export interface TherapyMeasurementCreate {
  value: number;
  notes?: string | null;
  session_id?: string | null;
  measured_at: string;
}

export const indicatorsApi = {
  list: (patientId: string): Promise<TherapyIndicator[]> =>
    request<TherapyIndicator[]>("GET", `/patients/${patientId}/indicators`),
  create: (patientId: string, body: TherapyIndicatorCreate): Promise<TherapyIndicator> =>
    request<TherapyIndicator>("POST", `/patients/${patientId}/indicators`, body),
  get: (indicatorId: string): Promise<TherapyIndicatorWithMeasurements> =>
    request<TherapyIndicatorWithMeasurements>("GET", `/indicators/${indicatorId}`),
  update: (indicatorId: string, body: TherapyIndicatorUpdate): Promise<TherapyIndicator> =>
    request<TherapyIndicator>("PUT", `/indicators/${indicatorId}`, body),
  delete: (indicatorId: string): Promise<void> =>
    request<void>("DELETE", `/indicators/${indicatorId}`),
  addMeasurement: (indicatorId: string, body: TherapyMeasurementCreate): Promise<TherapyMeasurement> =>
    request<TherapyMeasurement>("POST", `/indicators/${indicatorId}/measurements`, body),
  listMeasurements: (indicatorId: string): Promise<TherapyMeasurement[]> =>
    request<TherapyMeasurement[]>("GET", `/indicators/${indicatorId}/measurements`),
};
